File: php/src/Application/Catalogue/CreateItemFamily.php
<?php

declare(strict_types=1);

namespace Alxarafe\App\Application\Catalogue;

use Alxarafe\App\Domain\Catalogue\Entity\ItemFamily;
use Alxarafe\App\Domain\Catalogue\ValueObject\ItemFamilyCode;
use Alxarafe\App\Domain\Catalogue\ValueObject\ItemFamilyId;
use Alxarafe\App\Domain\Catalogue\ValueObject\StorageAttributeId;

final readonly class CreateItemFamily
{
    public function __construct(private ItemFamilyRepository $repository)
    {
    }

    /** @param list<string> $attributeIds */
    public function execute(string $code, string $name, array $attributeIds): ItemFamily
    {
        $attributes = array_map(static fn (string $value): StorageAttributeId => new StorageAttributeId($value), $attributeIds);
        $family = new ItemFamily(ItemFamilyId::generate(), new ItemFamilyCode($code), $name, $attributes);
        $available = $this->repository->existingAttributeIds($attributes);
        foreach ($attributes as $attribute) {
            if (!in_array($attribute->value(), $available, true)) {
                throw new StorageAttributeNotFound('Storage attribute not found: ' . $attribute->value());
            }
        }

        if ($this->repository->findByCode($family->code()) !== null) {
            throw new ItemFamilyConflict('Item family code already exists.');
        }

        $this->repository->save($family);
        return $family;
    }
}

File: java/src/main/java/com/alxarafe/app/application/catalogue/CreateItemFamily.java
package com.alxarafe.app.application.catalogue;

import com.alxarafe.app.domain.catalogue.entity.ItemFamily;
import com.alxarafe.app.domain.catalogue.valueobject.ItemFamilyCode;
import com.alxarafe.app.domain.catalogue.valueobject.ItemFamilyId;
import com.alxarafe.app.domain.rules.valueobject.AttributeCode;

import java.util.List;

public final class CreateItemFamily {
    private final ItemFamilyRepository repository;

    public CreateItemFamily(ItemFamilyRepository repository) {
        this.repository = repository;
    }

    public ItemFamily execute(String code, String name, List<String> attributeCodes) {
        List<AttributeCode> attributes = attributeCodes.stream().map(AttributeCode::new).toList();
        ItemFamily family = new ItemFamily(ItemFamilyId.generate(), new ItemFamilyCode(code), name, attributes);
        List<String> available = repository.availableFamilyAttributes(attributes);
        for (AttributeCode attribute : attributes) {
            if (!available.contains(attribute.value())) {
                throw new IllegalArgumentException("Unknown FAMILY attribute: " + attribute.value());
            }
        }

        if (repository.findByCode(family.code()).isPresent()) {
            throw new ItemFamilyConflict("Item family code already exists.");
        }

        repository.save(family);
        return family;
    }
}
